cpmpy/solvers/pysat_rc2.py

```diff
@@ -0,0 +1,178 @@
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+
+class CPM_RC2(CPM_pysat):
+    """
+    Interface to PySAT's API
+
+    Requires that the 'python-sat' python package is installed:
+    $ pip install python-sat
+
+    See detailed installation instructions at:
+    https://pysathq.github.io/installation.html
+
+    Creates the following attributes (see parent constructor for more):
+    pysat_vpool: a pysat.formula.IDPool for the variable mapping
+    pysat_solver: a pysat.solver.Solver() (default: glucose4)
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import pysat
+            # there is actually a non-related 'pysat' package
+            # while we need the 'python-sat' package, some more checks:
+            from pysat.solvers import Solver
+            return True
+        except ImportError as e:
+            return False
+
+    @staticmethod
+    def solvernames():
+        """
+            Returns solvers supported by PySAT on your system
+        """
+        from pysat.solvers import SolverNames
+        names = []
+        for name, attr in vars(SolverNames).items():
+            if not name.startswith('__') and isinstance(attr, tuple):
+                if name not in attr:
+                    name = attr[-1]
+                names.append(name)
+        return names
+
+    def __init__(self, cpm_model=None, subsolver="glucose3"):
+        """
+        Constructor of the native RC2 solver object
+
+        Requires a CPMpy model as input, and will create the corresponding
+        PySAT clauses and RC2 solver object.
+
+        Only supports maximal satisfaction problems
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model(), optional
+        """
+        if not self.supported():
+            raise Exception("CPM_pysat: Install the python 'python-sat' package to use this solver interface (NOT the 'pysat' package!)")
+
+        from pysat.formula import WCNF
+        from pysat.examples.rc2 import RC2
+
+        assert subsolver in CPM_RC2.solvernames(), f"Wrong solver ({subsolver}) selected from available: ({CPM_RC2.solvernames()})"
+
+        self.pysat_solver = RC2(formula=WCNF(), solver=subsolver)
+        self.pysat_vpool = self.pysat_solver.pool
+
+        SolverInterface.__init__(self, name="rc2", cpm_model=cpm_model)
+
+    def solve(self, assumptions=None):
+        """
+            Call the MaxSAT RC2 solver
+
+            Arguments:
+            - assumptions: list of CPMpy Boolean variables that are assumed to be true.
+                           MaxSAT solver does not support incremental solving and assumptions.
+                           CPM_rc2 adds the assumptions as hard clauses to the MaxSAT solver.
+        """
+        if assumptions is not None:
+            print("[Warning] Assumption variables interpreted as hard clauses!")
+            pysat_assum_vars = self.solver_vars(assumptions)
```

Comment:
Can you extract the clauses from the pysat model? If so, I would re-initialize the RC2 solver to remove assumptions after solving. This way the semantics are consistent with other solvers. As the constraints do not have to pass by our transformations again, this should be quite efficient

Reply:
Absolutely, I will keep track of the clauses and save them into a WCNF formula that can be reinitialized at every solve call with no assumptions.